src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -596,17 +610,73 @@ public override void Dispose()
 
         private void Dispose(bool disposing)
         {
-            if (_disposed)
+            int disposed = Interlocked.Exchange(ref _disposed, 1);
+            if (disposed != 0)
             {
                 return;
             }
 
-            _disposed = true;
-            _state.Handle.Dispose();
-            Marshal.FreeHGlobal(_state.SendQuicBuffers);
-            if (_stateHandle.IsAllocated) _stateHandle.Free();
-            CleanupSendState(_state);
-            _state.ConnectionState?.RemoveStream(this);
+            bool callShutdown = false;
+            bool abortRead = false;
+            bool releaseHandles = false;
+            lock (_state)
+            {
+                if (_canWrite && _state.ShutdownWriteState == ShutdownWriteState.None)
+                {
+                    callShutdown = true;
+                }
+
+                if (_canRead && _state.ReadState != ReadState.Aborted && _state.ReadState != ReadState.ReadsCompleted)
+                {
+                    abortRead = true;
+                    _state.ReadState = ReadState.Aborted;
+                }
+
+                if (_state.ShutdownState != ShutdownState.Finished)
+                {
+                    _state.ShutdownState = ShutdownState.Pending;
+                }
+                else if (!callShutdown && !abortRead)
+                {
+                    _state.ShutdownState = ShutdownState.Finished;
+                }
+
+                // We already got final event.
+                releaseHandles = Interlocked.Exchange(ref _state.ShutdownDone, 1) == 2;
```

Comment:
If we already got a final event, then do we actually need to call shutdown/abort below?

Reply:
The way how it currently work is that ShutdownComplete will not release everything if Dispose was not called. See note bellow. We could change that but then from example Shutdown and/or Rest can throw if handle is gone.